Fix #15571 panic on write to source parquet file

jjflash95 opened the pull request:
Fixes #15571 

# Description

Writing to a source `.parquet` (`polars open some_file.parquet | polars save some_file.parquet`) file made the plugin panic, added a guard to check the data_source path as per [this comment](https://github.com/nushell/nushell/issues/15571#issuecomment-2812707161)

Example output now:
<img width="850" alt="Screenshot 2025-04-18 at 21 10 30" src="https://github.com/user-attachments/assets/40a73cc7-6635-43dc-a423-19c7a0c8f59c" />

# User-Facing Changes
<!-- List of all changes that impact the user experience here. This helps us keep track of breaking changes. -->

# Tests + Formatting
 - Add 1 test
 - clippy OK
 - cargo fmt OK

# After Submitting
No action required

jjflash95 commented on crates/nu_plugin_polars/src/dataframe/command/core/save/mod.rs:
```diff
@@ -112,11 +112,20 @@ impl PluginCommand for SaveDF {
+        let metadata = input.metadata();
         let value = input.into_value(call.head)?;
 
+        check_writing_into_source_file(
```
this is a slightly modified version of [filesystem::open implementation](https://github.com/nushell/nushell/blob/main/crates/nu-command/src/filesystem/save.rs#L289)

jjflash95 commented on crates/nu_plugin_polars/src/dataframe/command/core/save/mod.rs:
```diff
@@ -223,6 +230,28 @@ fn command(
     Ok(PipelineData::empty())
 }
 
+fn check_writing_into_source_file(
```
this is a slightly modified version of [filesystem::save::check_saving_to_source_file](https://github.com/nushell/nushell/blob/main/crates/nu-command/src/filesystem/save.rs#L289)

jjflash95 commented on crates/nu_plugin_polars/src/dataframe/command/core/save/mod.rs:
```diff
@@ -223,6 +230,28 @@ fn command(
+    let Some(DataSource::FilePath(source)) = metadata.map(|meta| &meta.data_source) else {
+        return Ok(());
+    };
+
```
Can be changed to only guard `.parquet` files, since this panic doesn't happen on `.csv` ones for example, but I'm not super familiar with polars to make this call, so for consistency I'm including all extensions

sholderbach replied:
I am not sure if despite not panicking the reader may read an empty csv as the writer starts opening the file and thus corrupting it.

jjflash95 replied:
Just tested and it indeed is silently corrupting the csv, so I guess this is even better because it will prevent silent errors

sholderbach commented on crates/nu_plugin_polars/src/dataframe/command/core/open.rs:
```diff
@@ -164,6 +164,8 @@ fn command(
 
     let hive_options = build_hive_options(plugin, call)?;
+    let metadata = PipelineMetadata::default()
+        .with_data_source(DataSource::FilePath(spanned_file.item.clone().into()));
```
`polars open` supports additionally to load things directly from a cloud provider. So in that case the file path wouldn't point to a local filepath but a URI. 

In principle there shouldn't be an issue of adding it to the `DataSource::FilePath` as well, so we would not have to necessarily set this only when hitting the local FS. The only concern I have is if we have any readers of `DataSource::FilePath` that assume a local file system and e.g. path normalization on Windows suddenly pukes on the URL.

ayax79 replied:
I could see that allowing URLs in the file path could be a useful thing in general, but it could be a big change. Most of the polars commands don't currently maintain the metadata, so it will get lost pretty much immediately right now. It is something I should fix.

ayax79 replied:
I have confirmed that reading parquet and csv files still works from AWS.

jjflash95 replied:
is it worth adding a new `DataSource::External(URI)` member to the enum, or should I just not add the metadata when cloud_options are present?

sholderbach replied:
> Most of the polars commands don't currently maintain the metadata, so it will get lost pretty much immediately right now

To truly fix #15571 using the metadata trick we would probably need to tweak/check all polars command that they propagate the metadata like the core nushell filter commands. 

There may be alternate ways to implement this protection, but then you woud also need to account for false positives, when there are steps collecting into memory in between where further protection is not necessary.

jjflash95 replied:
Ended up adding the new `DataSource`, as for fixing the issue fully I think that can be done in a separate PR/issue?

ayax79 replied:
I would prefer the rest of the command fixes happen in a separate PR.

sholderbach reviewed (commented):
Thanks for implementing that, I would pull in @ayax79  briefly if he has anything to add on the polars and object store specific stuff.

jjflash95 commented on crates/nu-command/src/debug/util.rs:
```diff
@@ -0,0 +1,35 @@
+        content_type,
+    }) = metadata
+    {
+        match data_source {
```
extracted method from 3 implementations, changed to exhaustive pattern match so if anyone adds a new DataSource they don't bang their heads against the wall wondering why `some_command_with_new_data_src | metadata` comes out empty

sholderbach commented on crates/nu-protocol/src/pipeline/metadata.rs:
```diff
@@ -34,6 +35,7 @@ pub enum DataSource {
     Ls,
     HtmlThemes,
     FilePath(PathBuf),
+    Url(Box<Url>),
```
Heretical question, does it need to be a `Url` type? All uses we have so far just treat it as a string and we would be paying the small cost for the `serde` implementation.

jjflash95 replied:
Doesn't really need to be `Url` I kinda auto-piloted towards that without putting it much thought. Changed to be a string

fdncred replied:
I'm not sure we need a Url and a Uri. We should probably pick one. https://github.com/nushell/nushell/pull/15649

jjflash95 replied:
Yes, you are right, my changes are from a few days before that PR, so have I known they were planning on adding the extra datasource I wouldn't have done it.

Either way, @ayax79 suggested below that URL should not be an extra datasource, so I think we are going to scratch both implementations off, mine and https://github.com/nushell/nushell/pull/15649

sholderbach commented on crates/nu_plugin_polars/src/dataframe/command/core/open.rs:
```diff
@@ -165,6 +166,17 @@ fn command(
     let hive_options = build_hive_options(plugin, call)?;
 
+    let uri = spanned_file.item.clone();
+    let data_source = if resource.cloud_options.is_some() {
+        Url::from_str(&uri)
+            .map(|url| DataSource::Url(url.into()))
+            .unwrap_or_else(|_| DataSource::FilePath(uri.into()))
```
I am not as familiar as @ayax79 if there may be other paths or if `cloud_options` is the clear indicator on whether it is a local or uri path.

ayax79 replied:
Currently, cloud options is created if it is a supported cloud uri... right now only s3://. If it isn't an s3 resource, then cloud_options will be None. This should be fine. Really, we could support Url across the board at some point and provide a file:// url.. I wouldn't worry about it for now.

I'll be adding to other cloud providers at some point.

sholderbach reviewed (commented):
Thanks for iterating, respect for picking the hardest route.

Adding it to `metadata set` can be done in a follow up PR.

Getting the commands to pass through metadata would be higher priority.

If that's not done the title `Fix #15571...` is still a bit aspirational

jjflash95 commented:
Added metadata propagation to *almost* all polars commands.

Attached below is the list of files that were not adjusted to propagate input metadata, all others were changed to do it.

One thing that is not clear to me is how to handle this with commands that take from more than one data source, for example `join`
```
crates/nu_plugin_polars/src/dataframe/command/core/open.rs

crates/nu_plugin_polars/src/dataframe/command/aggregation/mod.rs
crates/nu_plugin_polars/src/dataframe/command/boolean/mod.rs
crates/nu_plugin_polars/src/dataframe/command/core/mod.rs
crates/nu_plugin_polars/src/dataframe/command/core/resource.rs
crates/nu_plugin_polars/src/dataframe/command/core/save/arrow.rs
crates/nu_plugin_polars/src/dataframe/command/core/save/avro.rs
crates/nu_plugin_polars/src/dataframe/command/core/save/csv.rs
crates/nu_plugin_polars/src/dataframe/command/core/save/mod.rs
crates/nu_plugin_polars/src/dataframe/command/core/save/ndjson.rs
crates/nu_plugin_polars/src/dataframe/command/core/save/parquet.rs
crates/nu_plugin_polars/src/dataframe/command/core/schema.rs
crates/nu_plugin_polars/src/dataframe/command/core/shape.rs
crates/nu_plugin_polars/src/dataframe/command/core/to_dtype.rs
crates/nu_plugin_polars/src/dataframe/command/core/to_nu.rs
crates/nu_plugin_polars/src/dataframe/command/core/to_repr.rs
crates/nu_plugin_polars/src/dataframe/command/core/to_schema.rs
crates/nu_plugin_polars/src/dataframe/command/data/append.rs
crates/nu_plugin_polars/src/dataframe/command/data/collect.rs
crates/nu_plugin_polars/src/dataframe/command/data/dummies.rs
crates/nu_plugin_polars/src/dataframe/command/data/lit.rs
crates/nu_plugin_polars/src/dataframe/command/data/mod.rs
crates/nu_plugin_polars/src/dataframe/command/data/sample.rs
crates/nu_plugin_polars/src/dataframe/command/data/sql_context.rs
crates/nu_plugin_polars/src/dataframe/command/data/sql_expr.rs
crates/nu_plugin_polars/src/dataframe/command/datetime/mod.rs
crates/nu_plugin_polars/src/dataframe/command/index/arg_max.rs
crates/nu_plugin_polars/src/dataframe/command/index/arg_min.rs
crates/nu_plugin_polars/src/dataframe/command/index/arg_sort.rs
crates/nu_plugin_polars/src/dataframe/command/index/arg_unique.rs
crates/nu_plugin_polars/src/dataframe/command/index/mod.rs
crates/nu_plugin_polars/src/dataframe/command/index/set_with_idx.rs
crates/nu_plugin_polars/src/dataframe/command/integer/mod.rs
crates/nu_plugin_polars/src/dataframe/command/list/mod.rs
crates/nu_plugin_polars/src/dataframe/command/mod.rs
crates/nu_plugin_polars/src/dataframe/command/string/mod.rs
crates/nu_plugin_polars/src/dataframe/command/stub.rs
```

jjflash95 commented:
Basically what I did was change commands that did not `collect` the dataframe to propagate the metadata.
As a follow up we could also discriminate from eager and lazy dataframes using content-type, to avoid erroring when writing to source for --eager data (that is already loaded into memory).

The commands that were not changed will overwrite the metadata, which I think is what we intend here


Example of commands propagating metadata:
<img width="1107" alt="Screenshot 2025-04-25 at 2 52 32 AM" src="https://github.com/user-attachments/assets/820689f9-5c6b-4465-91f9-df90aa69e0fb" />

Adding a non-propagating command at the end will override the metadata
<img width="1193" alt="Screenshot 2025-04-25 at 2 55 12 AM" src="https://github.com/user-attachments/assets/4302da4c-8951-4b04-9b69-472a8c860cec" />

jjflash95 commented:
> I would prefer the rest of the command fixes happen in a separate PR.

Just read this. If it helps I can cherry-pick the command changes onto a separate PR and only leave the logic relevant to `DataSource` in this one

ayax79 commented:
@jjflash95, if you already included them we can keep them in this one. We will merge this in a few days. There is a release tomorrow and we avoiding merging PRs around the release.

ayax79 commented:
Also, per discussion at [#15649](https://github.com/nushell/nushell/pull/15649) I don't think that URL should really be a datasource, but rather extra metadata. Much the way that content type is.

jjflash95 commented:
Do you think it is worth renaming `DataSource::FilePath` to `DataSource::Uri`? I think `FilePath` might be misleading if we want to include all possible resource schemes.

As for the extra metadata, do you think it is worth it being an enum? I'm just thinking if we may add more schemes down the line like S3 urls (i.e. CloudOptions) this could be helpful, but I'm not completely sure if it is justified for our current state

```
pub enum Scheme {
    #[default]
    None,
    Http,
    Https,
    File,
}

pub struct PipelineMetadata {
    pub data_source: DataSource,
    pub content_type: Option<String>,
    pub scheme: Scheme,
}

```

ayax79 commented:
Hi @jjflash95, The core team met to discuss this ticket. The group decided that we did not want to introduce URL metadata at this time.

This involved discussing on whether it should be a DataSource or separate metadata and the usefulness across the board. For instance, DataSource doesn't really fit as even file paths could be represented as URLs. If we introduced additional metadata, where else would it really be utilized.

jjflash95 commented:
Hi @ayax79 makes sense, I had the hunch that it was a bit of overkill adding the new DataSource but don't have enough experience to make that call.

Anyways I updated the PR and removed all references to DataSource::Url, hope this is enough to close the issue!

jjflash95 commented:
Just a small nudge to ensure this does not get buried in the PR list.

I think this meets the merging criteria, also this PR being so big is just waiting for conflicts to arise.

cc: @sholderbach (sorry for the direct tag)

jjflash95 commented:
rebased from main and resolved conflicts.

fdncred commented:
Does this PR have anything to do with the title? Does it still help with the panic?

It seems like this PR now is just a refactor of the metadata functionality.

jjflash95 commented:
This PR still does what the title says, which is preventing panics when writing into a source file, to accomplish this there are two main changes:
 - Check the metadata for a data source and ensure its not the same as the target path when saving a polars data frame
 - Propagating the metadata throughout the relevant polars commands (this ensures that the metadata is not lost when chaining multiple commands before polars save)

fdncred commented:
ok, thanks for explaining.

ayax79 reviewed (changes requested):
see my feedback on the run_inner stuff. It adds too much unnecessary boiler plate.

jjflash95 commented on crates/nu_plugin_polars/src/dataframe/command/aggregation/mean.rs:
```diff
@@ -89,6 +89,20 @@ impl PluginCommand for ExprMean {
+}
+
+impl ExprMean {
+    fn run_inner(
```
I totally missed this comment, the ‘run_inner’ issue was because I just ran a macro to edit the +100 files and this was the only way I found to generalize the impl to work on all these files.

I’ll try to change this during next week

jjflash95 commented:
Comments addressed!

jjflash95 commented:
@ayax79 PTAL when you can
